File: ios/Classes/Tasks/LocationTask.m
//
//  LocationTask.m
//  geolocator
//
//  Created by Maurits van Beusekom on 20/05/2019.
//

#import "LocationTask.h"


@implementation LocationTask {
    LocationOptions *_locationOptions;
}

- (void)dealloc {
    NSLog(@"[LocationTask %s]", sel_getName(_cmd));
}

- (instancetype)initWithContext:(TaskContext *)context completionHandler:(CompletionHandler)completionHandler {
    self = [super initWithContext:context completionHandler:completionHandler];
    if (self) {
        _locationOptions = [[LocationOptions alloc] initWithArguments:context.arguments];
    }
    
    return self;
}

- (void)startTask {
    if (_locationManager == nil) {
        _locationManager = [[CLLocationManager alloc] init];
        _locationManager.delegate = self;
    }
    
    CLLocationAccuracy accuracy = [self clValue:_locationOptions.accuracy];
    CLLocationDistance distanceFilter = _locationOptions.distanceFilter;
    
    _locationManager.desiredAccuracy = accuracy;
    _locationManager.distanceFilter = distanceFilter;
}

- (void)stopTask {
    if (_locationManager != nil) {
        [_locationManager stopUpdatingLocation];
        _locationManager = nil;
    }
    
    [super stopTask];
}

- (void)locationManager:(CLLocationManager *)manager  didUpdateLocations:(NSArray<CLLocation *> *)locations {
    NSLog(@"%s", sel_getName(_cmd));
    if ([locations lastObject]) {
        CLLocation *location = [locations lastObject];
        NSDictionary *positionDict = [location toDictionary];
        
        [[self context] resultHandler](positionDict);
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    NSLog(@"%s", sel_getName(_cmd));
    if([error.domain isEqualToString:kCLErrorDomain] && error.code == kCLErrorLocationUnknown) {
        return;
    }
    
    [[self context] resultHandler]([FlutterError errorWithCode:@"ERROR_UPDATING_LOCATION" message:error.localizedDescription details:@(error.code)]);
    [self stopTask];
}

- (CLLocationAccuracy)clValue:(GeolocationAccuracy)accuracy {
    switch (accuracy) {
        case GeolocationAccuracyLowest:
            return kCLLocationAccuracyThreeKilometers;
        case GeolocationAccuracyLow:
            return kCLLocationAccuracyKilometer;
        case GeolocationAccuracyMedium:
            return kCLLocationAccuracyHundredMeters;
        case GeolocationAccuracyHigh:
            return kCLLocationAccuracyNearestTenMeters;
        case GeolocationAccuracyBest:
            return kCLLocationAccuracyBest;
        case GeolocationAccuracyBestForNavigation:
            return kCLLocationAccuracyBestForNavigation;
        default:
            return 0.0;
    }
}
@end


@implementation CurrentLocationTask

- (void)dealloc {
    NSLog(@"[CurrentLocationTask %s]", sel_getName(_cmd));
}

- (void)startTask {
    [super startTask];
    
    [_locationManager requestWhenInUseAuthorization];
    
    if (@available(iOS 9.0, *)) {
        [_locationManager requestLocation];
    } else {
        [_locationManager startUpdatingLocation];
    }
}
	
- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    [super locationManager:manager didUpdateLocations:locations];
    [self stopTask];
}

@end


@implementation StreamLocationUpdatesTask

- (void)dealloc {
    NSLog(@"%s", sel_getName(_cmd));
}

- (void)startTask {
    [super startTask];
    
    [_locationManager startUpdatingLocation];
}

@end

File: ios/Classes/Tasks/LocationTask.swift
//
//  LocationServices.swift
//  geolocator
//
//  Created by Maurits van Beusekom on 30/06/2018.
//

import CoreLocation
import Flutter
import Foundation

class LocationTask : Task, TaskProtocol, CLLocationManagerDelegate {
    private var _locationOptions: LocationOptions
    private var _locationManager: CLLocationManager?
    
    required init(
        context: TaskContext,
        completionHandler: CompletionHandler?) {
        
        self._locationOptions = Codec.decodeLocationOptions(from: context.arguments)
        
        super.init(context: context,
                   completionHandler: completionHandler)
    }
    
    func startTask() {
        if(_locationManager == nil)
        {
            _locationManager = CLLocationManager()
            _locationManager!.delegate = self
        }

        _locationManager!.desiredAccuracy = _locationOptions.accuracy.clValue;
        _locationManager!.distanceFilter = _locationOptions.distanceFilter;
        _locationManager!.startUpdatingLocation()
        _locationManager!.startUpdatingHeading()
    }
    
    override func stopTask() {
        if(_locationManager != nil) {
            _locationManager!.stopUpdatingLocation()
            _locationManager!.stopUpdatingHeading()
            _locationManager = nil
        }
        
        super.stopTask()
    }
}

final class CurrentLocationTask : LocationTask {
    private var lastLocation : CLLocation?
    private var lastHeading : CLHeading?
    
    func locationManager(_ manager: CLLocationManager, didUpdateHeading heading: CLHeading) {
        lastHeading = heading
        complete()
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        lastLocation = locations.last
        complete()
    }
    
    func complete(){
        if (lastLocation != nil && lastHeading != nil){
            let positionDict = lastLocation!.toDictionary(heading: lastHeading!.trueHeading)
            context.resultHandler(positionDict)
            stopTask()
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        context.resultHandler(FlutterError.init(
            code: "ERROR_UPDATING_LOCATION",
            message: error.localizedDescription,
            details: nil))
        
        stopTask()
    }
}

final class StreamLocationUpdatesTask : LocationTask {
    private var lastHeading : CLHeading?
    
    func locationManager(_ manager: CLLocationManager, didUpdateHeading heading: CLHeading) {
        lastHeading = heading
    }
    
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        
        if (lastHeading != nil){
            let positionDict = location.toDictionary(heading: lastHeading!.trueHeading)
            context.resultHandler(positionDict)
        } else {
            let positionDict = location.toDictionary(heading: 0.0)
            context.resultHandler(positionDict)
        }
    
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        context.resultHandler(FlutterError.init(
            code: "ERROR_UPDATING_LOCATION",
            message: error.localizedDescription,
            details: nil))
        
        stopTask()
    }
}
